src/libraries/System.Runtime/tests/System.Globalization.Tests/CompareInfo/CompareInfoTests.HashCode.cs

```diff
@@ -12,108 +12,52 @@ namespace System.Globalization.Tests
-
-            if (areEqual)
-            {
-                Assert.True(areHashCodesEqual);
-            }
-            else
+            int additionalCollisions = 0;
+            CultureInfo[] cultures = CultureInfo.GetCultures(CultureTypes.AllCultures);
+            foreach (CultureInfo culture in cultures)
             {
-                Assert.False(areHashCodesEqual);
-            }
-
-            // implication of the above behavior:
-            StringComparer stringComparer = new CustomComparer(cmpInfo, options);
-            TryAddToCustomDictionary(stringComparer, str1, str2, areHashCodesEqual);
-        }
+                // Japanese does not have "None" compare option, it always ignores Kana
+                // HashCode is not available for options different than None or IgnoreCase
+                if (culture.Name.Split('-')[0] == "ja")
+                    continue;
 
-        private void TryAddToCustomDictionary(StringComparer comparer, string str1, string str2, bool shouldFail)
-        {
-            Dictionary<string, int> customDictionary = new Dictionary<string, int>(comparer);
-            customDictionary.Add(str1, 0);
-            try
-            {
-                customDictionary.Add(str2, 1);
-                Assert.False(shouldFail);
-            }
-            catch (ArgumentException ex)
-            {
-                Assert.True(shouldFail);
-                Assert.Contains("An item with the same key has already been added.", ex.Message);
+                for (int i = 0; i <= 65535; i++)
+                    CheckChar(i, culture);
             }
-            catch (Exception ex)
+
+            void CheckChar(int charCode, CultureInfo culture)
             {
-                Assert.Fail($"Unexpected exception thrown: {ex}");
+                var cmpInfo = culture.CompareInfo;
+                char character = (char)charCode;
+                string str1 = $"a{character}b";
+                string str2 = "ab";
+                CompareOptions options = CompareOptions.None;
+                var hashCode1 = cmpInfo.GetHashCode(str1, options);
+                var hashCode2 = cmpInfo.GetHashCode(str2, options);
+                bool areHashCodesEqual = hashCode1 == hashCode2;
+                StringComparer stringComparer = new CustomComparer(cmpInfo, options);
+                // if equal => same, then expect hash => same  
+                if (stringComparer.Compare(str1, str2) == 0)
+                {
+                    Assert.True(areHashCodesEqual, $"Expected equal hashes for equal strings. The check failed for culture {culture.Name}, character: {character}, code: {charCode}.");
+                }
+                // if equal => diff, then expect hash => diff
+                else
+                {
+                    if (areHashCodesEqual)
+                    {
+                        additionalCollisions++; // this should be smallest possible, 11541466
```

Comment:
Should we add some assert to check if `additionalCollisions` grows over certain limit? Or maybe some logging of the value so that we can keep track if the collisions became to often? Otherwise I'm not really sure we need this else branch in the `CheckChar` function.

Reply:
It's already very high, 74%. I am not sure what would be the "limit". It looks that it has no current use but I wanted to save it here so that in re-checking the method (e.g when the previous assert fails) it would be easier to understand what's the mechanism and what was the state during writing the test.